shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Value.hs

```diff
@@ -0,0 +1,407 @@
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Shelley.Spec.Ledger.Value
+ where
+
+import           Cardano.Binary (ToCBOR, FromCBOR, toCBOR, fromCBOR, encodeListLen,
+                  decodeWord)
+import           Cardano.Prelude (NoUnexpectedThunks(..))
+import           Data.Coerce (coerce)
+
+import           Shelley.Spec.Ledger.BaseTypes (invalidKey)
+import           Shelley.Spec.Ledger.Coin (Coin (..))
+import           GHC.Generics (Generic)
+import           Data.Word (Word8)
+import           Cardano.Crypto.Hash (Hash, ShortHash)
+--import           Data.Group (Group (..))
+--import           Data.Semigroup
+--import           Data.Monoid
+-- import           Algebra.Module
+import           Data.Map.Strict (Map, elems, empty, toList, filterWithKey, keys,
+                 toList, singleton, lookup, insert, unionWith)
+import           Shelley.Spec.Ledger.Crypto
+import           Data.ByteString.Char8 (ByteString, pack) -- TODO is this the right Bytestring
+import           Shelley.Spec.Ledger.Scripts
+
+
+
+
+-- | Quantity
+newtype Quantity = Quantity Integer
+  deriving (Show, Eq, Generic, ToCBOR, FromCBOR, Ord, Integral, Num, Real, Enum, NoUnexpectedThunks)
+
+-- | Asset ID
+newtype AssetID = AssetID ByteString
+  deriving (Show, Eq, ToCBOR, FromCBOR, Ord, NoUnexpectedThunks)
+
+assetID :: AssetID -> ByteString
+assetID (AssetID aid) = aid
+
+-- | Policy ID
+newtype PolicyID crypto = PolicyID (ScriptHash crypto)
+  deriving (Show, Eq, ToCBOR, FromCBOR, Ord, NoUnexpectedThunks)
+
+policyID :: PolicyID crypto -> ScriptHash crypto
+policyID (PolicyID pid) = pid
+
+-- | A cryptocurrency value. This is a map from 'PolicyID's to a
+-- quantity of assets with this policy.
+--
+-- Operations on assets are usually implemented /pointwise/. That is,
+-- we apply the operation to the quantities for each asset in turn. So
+-- when we add two 'Value's the resulting 'Value' has, for each asset,
+-- the sum of the quantities of /that particular/ asset in the argument
+-- 'Value'. The effect of this is that the assets in the 'Value' are "independent",
+-- and are operated on separately.
+--
+-- Whenever we need to get the quantity of an asset in a 'Value' where there
+-- is no explicit quantity of that asset in the 'Value', then the quantity is
+-- taken to be zero.
```

Comment:
I still think that the best way to explain this is that they're finitely supported functions. Then the behaviour is obvious, rather than this being what sounds like a special case.